File: cpp/nucleotide-count/nucleotide_count.cpp
#include "nucleotide_count.h"

// using namespace std;

namespace nucleotide_count
{
    counter::counter(const std::string &dna)
    {
        for (auto nucleotide : dna)
        {
            if (nucleotides.count(nucleotide) == 0)
            {
                throw std::invalid_argument("Invalid nucleotide");
            }

            nucleotides[nucleotide]++;
        }
    }

    int counter::count(const char &nucleotide) const
    {
        if (nucleotides.count(nucleotide) == 0)
        {
            throw std::invalid_argument("Invalid nucleotide");
        }

        return nucleotides.at(nucleotide);
    }

    std::map<char, int> counter::nucleotide_counts() const
    {
        return nucleotides;
    }
} // namespace nucleotide_count

File: ruby/nucleotide-count/nucleotide_count.rb
class Dna
  def initialize(raw_dna)
    @nucleotides = {
      "A" => 0,
      "T" => 0,
      "C" => 0,
      "G" => 0,
    }

    raw_dna.each_char do |c|
      raise ArgumentError unless @nucleotides.has_key?(c.upcase)

      @nucleotides[c.upcase] += 1
    end
  end

  def histogram
    @nucleotides
  end

  def count(nucleotide)
    @nucleotides[nucleotide]
  end
end

module Nucleotide
  def self.from_dna(raw_dna)
    Dna.new(raw_dna)
  end
end
